### PR TITLE
Bug #3061: Fix converse header rank race condition in list send

### PR DESCRIPTION
**Pull request by rbuch**

### CODE REVIEW

**Review by stwhite91 — APPROVED** [review]

File: src/conv-core/convcore.C [hunk 1]
```diff
@@ -3031,12 +3031,32 @@ void CmiGroupInit(void)
...
+  // for local PEs and there is a race between setting the rank in the Converse header and
+  // the actual comm thread send for remote PEs
+#if CMK_SMP
+  for (int i = 0; i < npes - 1; i++)
+  {
+    const char* msgCopy = CmiCopyMsg(msg, len);
+    CmiSyncSendAndFree(pes[i], len, msgCopy);
   }
+
+  if (npes > 0) CmiSyncSend(pes[npes - 1], len, msg);
```
Comment (matthiasdiener) [inline src/conv-core/convcore.C:3046]:
```suggestion
  if (npes > 0) 
    CmiSyncSend(pes[npes - 1], len, msg);
```

File: tests/charm++/megatest/listsend.C [hunk 2]
```diff
@@ -0,0 +1,48 @@
+#include "listsend.h"
+#include <numeric>
+
+#define VALUE 0x01ABCDEF
+
+void listsend_moduleinit(void) {}
```
Comment (matthiasdiener) [inline tests/charm++/megatest/listsend.C:6]: no need for `(void)` as the argument for these functions.

File: src/conv-core/convcore.C (hunk 1 above)
Comment (matthiasdiener) [reply src/conv-core/convcore.C:3046]: Also, can you add a comment on why `npes-1` is treated separately?

**Review by matthiasdiener — APPROVED** [review]

File: src/conv-core/convcore.C [hunk 3]
```diff
@@ -3031,12 +3031,33 @@ void CmiGroupInit(void)
...
+    CmiSyncSendAndFree(pes[i], len, msgCopy);
   }
+  // No need to copy and free for the last send, so use the original message
+  if (npes > 0)
+    CmiSyncSend(pes[npes - 1], len, msg);
+#else
+  for (int i = 0; i < npes; i++)
+  {
+    // Copy if this is a self send to avoid unpack/send race
+    if (pes[i] == CmiMyPe() && npes > 1)
```
Comment (epmikida) [inline src/conv-core/convcore.C:3052]: Why would this cause a race? Wouldn't the unpack happen after I've finished doing all of the sends?

**Review by epmikida — APPROVED** [review]
LGTM pending the answer to one question

File: src/conv-core/convcore.C (hunk 3 above)
Comment (rbuch) [reply src/conv-core/convcore.C:3052]: Not necessarily, at least on some layers, even for non-SMP mode, the messaging layer still works by adding messages to a queue which it then processes between running EPs, but it checks its local queue before sending messages for external PEs (`src/arch/util/machine-common-core:1812-1816`):
```
    msg = CMIQueuePop(cs->recv); // checks local queue for msg
#endif
#if (!CMK_SMP || CMK_SMP_NO_COMMTHD) && !CMK_MULTICORE
    if (!msg) {
       AdvanceCommunication(0); // sends queued messages to remote PEs
```

When this happens, it is unpacked locally and then the same unpacked message is used for sends to remote PEs.

Comment (epmikida) [reply src/conv-core/convcore.C:3052]: Gotcha

**Pull request merged**